tests/common/helpers/parallel_utils.py

```diff
@@ -1,5 +1,209 @@
+        if not cls._instance:
+            with cls._lock:
+                if not cls._instance:
+                    cls._instance = super(InitialCheckState, cls).__new__(cls)
+                    cls._instance._initialize(num_followers, state_file)
+
+        return cls._instance
+
+    def _initialize(self, num_followers: int, state_file: str) -> None:
+        self.num_followers = num_followers
+        self.should_wait = num_followers > 0
+        self.state_file = state_file
+        self._set_initial_status()
+
+    def _set_initial_status(self) -> None:
+        with open(self.state_file, 'a+') as f:
+            fcntl.flock(f, fcntl.LOCK_EX)
+            f.seek(0)
+            content = f.readlines()
+            # only set the initial status if the file is empty
+            if not content:
+                f.write("{},{},{},{}\n".format(
+                    datetime.utcnow(),
+                    InitialCheckStatus.IDLE.value,
+                    0,
+                    ParallelRole.UNKNOWN.value,
+                ))
+
+                f.flush()
+            else:
+                logger.info("State file already contains data, skipping initial status setup")
+
+            fcntl.flock(f, fcntl.LOCK_UN)
+
+    def _read_state(self) -> Tuple[str, int]:
+        try:
+            content = read_last_line_of_file(self.state_file)
+            if not content:
+                logger.warning("State file is empty, returning IDLE status")
+                return InitialCheckStatus.IDLE.value, 0
+
+            _, status_value, acknowledgments, _ = content.split(',')
+            return status_value, int(acknowledgments)
+        except FileNotFoundError:
+            logger.warning("State file not found, returning IDLE status")
+            return InitialCheckStatus.IDLE.value, 0
+
+    def _is_expected_status(self, expected_status: InitialCheckStatus) -> bool:
+        status_value, _ = self._read_state()
+        if status_value == InitialCheckStatus.SANITY_CHECK_FAILED.value:
+            pt_assert(False, "Leader node sanity check failed. Exiting the test.")
+
+        return status_value == expected_status.value
+
+    def _acknowledge_status(self, ack_status: InitialCheckStatus, is_leader: bool) -> None:
+        with open(self.state_file, 'r+') as f:
+            fcntl.flock(f, fcntl.LOCK_EX)
+            lines = f.readlines()
+            if not lines:
+                raise Exception("State file is empty")
+
+            last_line = lines[-1].strip()
+            _, status_value, acknowledgments, _ = last_line.split(',')
+            if status_value != ack_status.value:
+                raise Exception("Cannot acknowledge status {} when status is {}".format(ack_status, status_value))
+
+            f.write("{},{},{},{}\n".format(
+                datetime.utcnow(),
+                ack_status.value,
+                int(acknowledgments) + 1,
+                ParallelRole.LEADER.value if is_leader else ParallelRole.FOLLOWER.value,
```

Comment:
Based on the discussion with Jianquan, we can also add the hostname here so we can debug it easily.